File: 0037-sudoku-solver/0037-sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m = len(board)
        n = len(board[0])
        rows = collections.defaultdict(set)
        cols = collections.defaultdict(set)
        square = collections.defaultdict(set)
        stack = []
        def backtracking(board):
            nonlocal rows
            nonlocal cols
            nonlocal square
            nonlocal stack
            if not stack:
                return True
            i, j = stack[-1]
            for num in range(1, 10):
                num_str = str(num)
                if num_str not in rows[i] and num_str not in cols[j] and num_str not in square[(i // 3, j // 3)]:
                    board[i][j] = num_str
                    rows[i].add(num_str)
                    cols[j].add(num_str)
                    square[(i // 3, j // 3)].add(num_str)
                    stack.pop()
                    if backtracking(board):
                        return True
                    board[i][j] = "."
                    rows[i].remove(num_str)
                    cols[j].remove(num_str)
                    square[(i // 3, j // 3)].remove(num_str)
                    stack.append((i, j))
            return False
                    


        for i in range(m):
            for j in range(n):
                if board[i][j] == ".":
                    stack.append((i, j))
                else:
                    rows[i].add(board[i][j])
                    cols[j].add(board[i][j])
                    square[(i // 3, j // 3)].add(board[i][j])
        

        backtracking(board)
        return board

        

File: 0037-sudoku-solver/0037-sudoku-solver.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        int m = board.size();
        int n = board[0].size();
        vector<unordered_set<char>> rows(m), cols(n), squares(9);
        vector<pair<int, int>> stack;

        // Lambda function for the backtracking
        function<bool(int)> backtracking = [&](int pos) -> bool {
            if (pos == stack.size()) return true;  // If there are no more positions on stack, we solved the puzzle

            auto [i, j] = stack[pos];
            int squareIndex = (i / 3) * 3 + (j / 3);
            
            for (char num = '1'; num <= '9'; ++num) {
                if (rows[i].count(num) || cols[j].count(num) || squares[squareIndex].count(num))
                    continue;  // Skip if number is already in the row, column, or square

                board[i][j] = num;
                rows[i].insert(num);
                cols[j].insert(num);
                squares[squareIndex].insert(num);

                if (backtracking(pos + 1)) return true;

                // Undo the move
                board[i][j] = '.';
                rows[i].erase(num);
                cols[j].erase(num);
                squares[squareIndex].erase(num);
            }
            return false;  // No suitable number was found
        };

        // Initialize sets and stack
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] == '.') {
                    stack.emplace_back(i, j);
                } else {
                    char num = board[i][j];
                    rows[i].insert(num);
                    cols[j].insert(num);
                    squares[(i / 3) * 3 + (j / 3)].insert(num);
                }
            }
        }

        backtracking(0);
    
    }
};